src/org/exist/xslt/TemplatesFactory.java

```diff
@@ -0,0 +1,151 @@
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.xml.transform.Templates;
+import javax.xml.transform.TransformerConfigurationException;
+import javax.xml.transform.sax.SAXTransformerFactory;
+import javax.xml.transform.sax.TemplatesHandler;
+import javax.xml.transform.sax.TransformerHandler;
+import org.exist.security.PermissionDeniedException;
+import org.exist.storage.DBBroker;
+import org.exist.xquery.Constants;
+import org.exist.xquery.value.NodeValue;
+import org.w3c.dom.Document;
+import org.xml.sax.SAXException;
+
+/**
+ * @author <a href="mailto:shabanovd@gmail.com">Dmitriy Shabanov</a>
+ */
+public class TemplatesFactory {
+
+  private final static ConcurrentMap<String, StylesheetUri> cache = new ConcurrentHashMap<>();
+
+  public static Stylesheet stylesheet(String stylesheet, String baseUri, Properties properties, boolean useCache) {
+
+    if (useCache && properties == null) {
+      return stylesheet(stylesheet, baseUri);
+    }
+
+    String uri = uri(stylesheet, baseUri);
+
+    return new StylesheetUri(uri, properties);
+  }
+
+  private static Stylesheet stylesheet(String stylesheet, String baseUri) {
+
+    String uri = uri(stylesheet, baseUri);
+
+    cache.computeIfAbsent(uri, StylesheetUri::new);
+    return cache.get(uri);
+  }
+
+  public static Stylesheet stylesheet(String stylesheet, String baseUri, boolean useCache) {
+    if (useCache) {
+      return stylesheet(stylesheet, baseUri);
+    } else {
+      return new StylesheetUri(uri(stylesheet, baseUri));
+    }
+  }
+
+  private static String uri(String stylesheet, String baseUri) {
+    String uri = stylesheet;
+    if (stylesheet.indexOf(':') == Constants.STRING_NOT_FOUND) {
+      Path f = Paths.get(stylesheet).normalize();
+      if (Files.isReadable(f)) {
+        uri = f.toUri().toASCIIString();
+      } else {
+        stylesheet = baseUri + File.separatorChar + stylesheet;
+        f = Paths.get(stylesheet).normalize();
+        if (Files.isReadable(f)) {
+          uri = f.toUri().toASCIIString();
+        }
+      }
+    }
+    return uri;
+  }
+
+  public static <E extends Exception> Stylesheet stylesheet(DBBroker broker, NodeValue node, String baseUri)
+      throws E, TransformerConfigurationException
+  {
+    SAXTransformerFactory factory = TransformerFactoryAllocator
+        .getTransformerFactory(broker.getBrokerPool());
+
+    String base = baseUri;
+    Document doc = node.getOwnerDocument();
+    if (doc != null) {
+      String uri = doc.getDocumentURI();
+
+      /*
+       * This must be checked because in the event the stylesheet is
+       * an in-memory document, it will cause an NPE
+       */
+      if (uri != null) {
+        base = uri.substring(0, uri.lastIndexOf('/'));
+      }
+    }
+
+    if (base != null) {
+      factory.setURIResolver(new EXistURIResolver(broker.getBrokerPool(), base));
+    }
+
+    return new Stylesheet() {
```

Comment:
It makes understanding the code harder. It is much nicer to have a clean separate subclass which has a javadoc description as to its purpose.